build-package.sh

```diff
@@ -672,6 +672,7 @@ for ((i=0; i<${#PACKAGE_LIST[@]}; i++)); do
 		cd "$TERMUX_PKG_MASSAGEDIR/$TERMUX_PREFIX_CLASSICAL"
 		termux_step_post_massage
 		cd "$TERMUX_PKG_MASSAGEDIR"
+		cp -a "./$TERMUX_PREFIX_CLASSICAL" "$TERMUX_BASE_DIR"
```

Comment:
> Correct me if I am wrong but I guess this is for on-device specific build issues?

This change is more relevant for compiling on non-devices, i.e. on a docker image.

> I don't think this is the best way to solve the type of problems you describe. This will add/modify files in $PREFIX that might not be known to package manager, and potentially create annoying-to-debug compilation errors.

The main "core" of the problem of unknown files for the package manager occurs when files are installed immediately on the system prefix after the files are compiled. Therefore, what I added can be called “flowers” :P

> though in ncurses case package should be marked as not safe for on-device builds in that case since it rm -Rf's stuff

It may sound nerding, but I personally think that we should not create packages that are dangerous to compile on devices (maybe I don’t understand something about such a solution).

Reply:
Are you sure build error was related to ncurse's post_massage step? I don't quite understand why build failed. From logs libedit failed to find curses.h (and friends), even though ncurses had been built before. 
Can you reproduce issue locally? Are the headers really missing from the prefix? 
What does libedit's config.log contain after failed build? Specifically, is it trying to find the headers in /data/data/com.termux/files/usr/include or in /data/data/com.termux/files/usr/glibc/include?

> It may sound nerding, but I personally think that we should not create packages that are dangerous to compile on devices (maybe I don’t understand something about such a solution).

I agree, we should work towards making all packages "safe" for building on device. One way is to make all packages install to TERMUX_PKG_MASSAGEDIR, as on build-on-device branch

Reply:
> Are you sure build error was related to ncurse's post_massage step? I don't quite understand why build failed. From logs libedit failed to find curses.h (and friends), even though ncurses had been built before.
Can you reproduce issue locally? Are the headers really missing from the prefix?
What does libedit's config.log contain after failed build?

I ran compilation of the `libedit` package on my computer without this commit, and got the same error. Here are the screenshots:
![IMG_20240604_200904_668](https://github.com/termux/termux-packages/assets/60917834/13c7166f-7398-4bb5-93ac-6cd433b47efa)
![IMG_20240604_200904_886](https://github.com/termux/termux-packages/assets/60917834/cd6b3e0c-872d-4ac7-acd0-c020455f647c)

> Specifically, is it trying to find the headers in /data/data/com.termux/files/usr/include or in /data/data/com.termux/files/usr/glibc/include?

It only finds headers in `/data/data/com.termux/files/usr/glibc/include`. It should not look for headers in the bionic environment (i.e. in `/data/data/com.termux/files/usr/include`), as this may break the compilation.

Reply:
So `/data/data/com.termux/files/usr/glibc/include/termcap.h` doesn't exist, so I guess `/data/data/com.termux/files/usr/glibc/include/ncursesw/termcap.h` is a dead symlink? 

Where was termcap.h installed to, the standard prefix? I guess that means that compiled ncurses-glibc does not have headers in the correct place?

Reply:
> So /data/data/com.termux/files/usr/glibc/include/termcap.h doesn't exist, so I guess /data/data/com.termux/files/usr/glibc/include/ncursesw/termcap.h is a dead symlink?

No, it's not a symbolic link, it's a file.

> Where was termcap.h installed to, the standard prefix? I guess that means that compiled ncurses-glibc does not have headers in the correct place?

All headers are installed in `/data/data/com.termux/files/usr/glibc/include/ncursesw`, but in the package all headers are located correctly after the massage step. Look at `build.sh` of the `ncurses` package to understand better - https://github.com/termux-pacman/glibc-packages/blob/main/gpkg/ncurses/build.sh

Reply:
Aha, so the ncurses build recipe in glibc-packages is quite different from termux-packages one.

Just move post_massage steps to post_make_install. Seems like termux-packages installs headers to several locations (include/, include/ncurses, include/ncursesw), while glibc-packages just installs to include/ncursesw for some reason.

post_massage should be used for creating empty folders, or removing duplicate files (though that could go in post_make_install as well), and other things that don't really change $PREFIX (as in termux-packages ncurses build recipe).

I think we can close this, this should preferably be fixed in glibc-packages ncurses build.sh

Reply:
> Seems like termux-packages installs headers to several locations (include/, include/ncurses, include/ncursesw), while glibc-packages just installs to include/ncursesw for some reason.

The `ncurses` package in termux-packages also uses `post_massage` to set headers in include/ncurses/ and in include/ncursesw/. They themselves are installed only in include/ during package compilation. I think this can also affect the compilation of the package that will try to find headers in include/ncurses/ or include/ncursesw/.

> post_massage should be used for creating empty folders, or removing duplicate files (though that could go in post_make_install as well), and other things that don't really change $PREFIX (as in termux-packages ncurses build recipe).

The post_massage is also used to create symbolic links or make minor changes to files. These are the kinds of changes that can affect the compilation of other packages, and such changes should be added to the system prefix.

Reply:
> The `ncurses` package in termux-packages also uses `post_massage` to set headers in include/ncurses/ and in include/ncursesw/. They themselves are installed only in include/ during package compilation. I think this can also affect the compilation of the package that will try to find headers in include/ncurses/ or include/ncursesw/.

Yeah, you are right, I was under the impression that it installed to multiple places and post_massage just replaced duplicate files with symlinks. Let's fix it by moving steps back to post_make_install. 

> The post_massage is also used to create symbolic links or make minor changes to files. These are the kinds of changes that can affect the compilation of other packages, and such changes should be added to the system prefix.

True, such steps should preferably go into post_make_install or similar as well